jgrapht-core/src/main/java/org/jgrapht/alg/tour/NearestNeighborHeuristicTSP.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * (C) Copyright 2019-2019, by Peter Harman and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.tour;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Random;
+import java.util.Set;
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.GraphTests;
+
+/**
+ * The nearest neighbour heuristic algorithm for the TSP problem.
+ *
+ * <p>
+ * The travelling salesman problem (TSP) asks the following question: "Given a
+ * list of cities and the distances between each pair of cities, what is the
+ * shortest possible route that visits each city exactly once and returns to the
+ * origin city?".
+ * </p>
+ *
+ * <p>
+ * This is perhaps the simplest and most straightforward TSP heuristic. The key
+ * to this algorithm is to always visit the nearest city.
+ * </p>
+ *
+ * <p>
+ * The implementation of this class is based on: <br>
+ * <a href="http://160592857366.free.fr/joe/ebooks/ShareData/Heuristics%20for%20the%20Traveling%20Salesman%20Problem%20By%20Christian%20Nillson.pdf">Nilsson
+ * C., "Heuristics for the Traveling Salesman Problem"</a>
+ * </p>
+ *
+ * <p>
+ * The runtime complexity of this class is $O(V^2)$.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Peter Harman
+ */
+public class NearestNeighborHeuristicTSP<V, E>
+        extends
+        HamiltonianCycleAlgorithmBase<V, E> {
+
+    private Random rng;
+    private V first;
+
+    /**
+     * Constructor. By default a random vertex is chosen to start.
+     */
+    public NearestNeighborHeuristicTSP() {
+        this(null, new Random());
+    }
+
+    /**
+     * Constructor
+     *
+     * @param first First vertex to visit, or bull to choose at random
```

Comment:
I think `bull` is a typo. I think you meant `null`?

Reply:
Oops, thanks!